drivers/osdp/src/osdp_pd.c

```diff
@@ -0,0 +1,576 @@
+	if (ret <= 0)
+		return 1;
+	if (new_packet && ret > 0)
+		pd->tstamp = millis_now();
+	pd->rx_buf_len += ret;
+
+#ifdef CONFIG_OSDP_PACKET_TRACE
+	if (pd->rx_buf[6] != CMD_POLL && pd->rx_buf[8] != CMD_POLL) {
+		osdp_dump("PD received", pd->rx_buf, pd->rx_buf_len);
+	}
+#endif
+
+	ret = phy_decode_packet(pd, pd->rx_buf, pd->rx_buf_len);
+	if (ret < 0) {
+		switch (ret) {
+		case -1: /* fatal error */
+			LOG_ERR(TAG "failed to decode packet");
+			pd->rx_buf_len = 0;
+			if (pd->channel.flush)
+				pd->channel.flush(pd->channel.data);
+			ret = -1;
+			break;
+		case -2: /* rx_buf_len != pkt->len; wait for more data */
+			ret = 1;
+			break;
+		case -3: /* soft fail - discard this message */
+		case -4: /* malformed OSDP packet - reset rx_buf_len */
+			pd->rx_buf_len = 0;
+			if (pd->channel.flush)
+				pd->channel.flush(pd->channel.data);
+			ret = 1;
+			break;
+		default:
+			ret = -1;
+		}
+		return ret;
+	}
+	pd->rx_buf_len = ret;
+	return 0;
+}
+
+void pd_state_update(struct osdp_pd *pd)
+{
+	int ret;
+	struct osdp_cmd reply;
+
+	switch (pd->state) {
+	case PD_STATE_IDLE:
+		ret = pd_receve_packet(pd);
+		if (ret < 0 || (pd->rx_buf_len > 0 &&
+		    millis_since(pd->tstamp) > OSDP_RESP_TOUT_MS)) {
+			/**
+			 * When we receive a command from PD after a timeout,
+			 * any established secure channel must be discarded.
+			 */
+			pd->state = PD_STATE_ERR;
+			break;
+		}
+		if (ret == 1)
+			break;
+		pd_decode_command(pd, &reply, pd->rx_buf, pd->rx_buf_len);
+		pd->state = PD_STATE_SEND_REPLY;
+		/* FALLTHRU */
+	case PD_STATE_SEND_REPLY:
+		if (pd_send_reply(pd, &reply) == -1) {
+			pd->state = PD_STATE_ERR;
+			break;
+		}
+		pd->rx_buf_len = 0;
+		pd->state = PD_STATE_IDLE;
+		break;
+	case PD_STATE_ERR:
+		/**
+		 * PD error state is momentary as it doesn't maintain any state
+		 * between commands. We just clean up secure channel status and
+		 * go back to idle state with a call to phy_state_reset().
+		 */
+		clear_flag(pd, PD_FLAG_SC_ACTIVE);
+		pd->rx_buf_len = 0;
+		pd->state = PD_STATE_IDLE;
+		break;
+	}
+}
+
+void osdp_pd_set_attributes(struct osdp_pd *pd, struct pd_cap *cap,
+			    struct pd_id *id)
+{
+	int fc;
+
+	while (cap && ((fc = cap->function_code) > 0)) {
+		if (fc >= CAP_SENTINEL)
+			break;
+		pd->cap[fc].function_code = cap->function_code;
+		pd->cap[fc].compliance_level = cap->compliance_level;
+		pd->cap[fc].num_items = cap->num_items;
+		cap++;
+	}
+
+	memcpy(&pd->id, id, sizeof(struct pd_id));
+}
+
+int osdp_pd_get_cmd(struct osdp_cmd *cmd)
+{
+	struct osdp *ctx = &osdp_ctx;
+	struct osdp_pd *pd = to_pd(ctx, 0);
+	struct osdp_cmd *f;
+
+	f = pd->queue.front;
+	if (f == NULL)
+		return -1;
+
+	memcpy(cmd, f, sizeof(struct osdp_cmd));
+	pd->queue.front = pd->queue.front->__next;
+	osdp_cmd_free(f);
+	return 0;
+}
+
+void osdp_refresh(void *arg1, void *arg2, void *arg3)
+{
+	struct osdp *ctx = &osdp_ctx;
+	struct osdp_pd *pd = to_pd(ctx, 0);
+
+	while (1) {
+		pd_state_update(pd);
+		k_sleep(K_MSEC(50));
```

Comment:
This is quite an active thread. Would it be possible to use a queue (fifo or lifo) instead so that one doesn't need to wake up every 50ms?

Reply:
I see. The thread basically is looking for incoming bytes and decoding them. I could make it so that the UART ISR wakes up this thread but is that what you had in mind?